generative_data_prep/processors/article_tokenizer.py

```diff
@@ -141,6 +148,200 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
+        """
+        if isinstance(jsonl, dict):
+            jsonl = [jsonl]
+
+        converted_jsonl = []
+        prompts = []
+        completions = []
+
+        """
+        TODO: Add a trick here to get the generation_prompt in tokenizer
+        This is a temporary solution, and the tokenizer should be modified in the future.
+        The tokenizer should be able to get the generation_prompt from the input text.
+        self.tokenizer.chat_template = "{% for message in messages %} {{'<|im_start|>' + message['role'] + '\n' +
+            message['content'] + '<|im_end|>' + '\n'}}{% endfor %}"
+        """
+
+        """
+        The mechanism designed here is to remember the original prompt and completion,
+        despite the transformation of the [prompt, completion] pair into a raw string
+          format with additional instruction tokens.
+
+        The following steps are performed:
+        1. Recognizes and preserves the original dialogue data format. (<prompt_placeholder>, <completion_placeholder>)
+        2. Converts the original [prompt, completion] format to the standard conversation [role, content] format.
+            {"prompt": "Hello, how are you?", "completion": "I'm doing great"} ->
+                {"role": "user", "content": "Hello, how are you?"}, {"role": "assistant", "content": "I'm doing great."}
+        """
+        for prompt_completion in jsonl:
+            if self.prompt_keyword not in prompt_completion:
+                err_msg = f"Prompt keyword required in every jsonl, {self.prompt_keyword} not found"
+                raise json.JSONDecodeError(err_msg, str(jsonl), 0)
+
+            if self.completion_keyword not in prompt_completion:
+                err_msg = f"Completion keyword required in every jsonl, {self.completion_keyword} not found"
+                raise json.JSONDecodeError(err_msg, str(jsonl), 0)
+
+            prompt = prompt_completion[self.prompt_keyword]
+            completion = prompt_completion[self.completion_keyword]
+            converted_jsonl.append({"role": "user", "content": DEFAULT_PROMPT_PLACEHOLDER})
+            converted_jsonl.append({"role": "assistant", "content": DEFAULT_COMPLETION_PLACEHOLDER})
+
+            prompts.append(prompt)
+            completions.append(completion)
+
+        if len(prompts) != len(completions):
+            err_msg = f"Number of prompts and completions must be equal, \
+             {len(prompts)} prompts and {len(completions)} completions found"
+            raise ValueError(err_msg)
+
+        if len(prompts) != len(jsonl):
+            err_msg = f"Number of prompts and jsonl must be equal, {len(prompts)} prompts and {len(jsonl)} jsonl found"
+            raise ValueError(err_msg)
+
+        # Apply specified chat templates using tokenizer (apply_chat_template) without tokenization.
+        formatted_chat = self.tokenizer.apply_chat_template(converted_jsonl, tokenize=False)
```

Comment:
Let's add an assertion here to make sure that tokenizer does have apply_chat_template defined?

Reply:
Thanks for pointing this out, I have revised the code in the latest commit.